src/main/java/com/google/devtools/build/lib/remote/RemoteDownloaderFactory.java

```diff
@@ -0,0 +1,43 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.remote;
+
+import com.google.common.base.Preconditions;
+import com.google.devtools.build.lib.bazel.repository.downloader.Downloader;
+import com.google.devtools.build.lib.bazel.repository.downloader.DownloaderFactory;
+import java.util.Optional;
+
+public final class RemoteDownloaderFactory implements DownloaderFactory {
+
+  private volatile Downloader remoteDownloader;
```

Comment:
Prefer synchronized over volatile. You almost never need it from a performance point of view, and it's much harder to reason about.

Reply:
This logic is copied from the existing delegate/factory stacks, such as `RepositoryRemoteExecutorFactoryDelegate`. To be honest I have no idea why this level of indirection is necessary, and would be happy to have some guidance (or, like, direct commits to this PR) from someone who writes Java regularly.

I tried using `synchronized` instead, but the build fails:
```
src/main/java/com/google/devtools/build/lib/remote/RemoteDownloaderFactory.java:24: error: modifier synchronized not allowed here
  private synchronized Downloader remoteDownloader;
```

Reply:
Oh, sorry. The synchronized goes on all methods that access the field, like so:

private Downloader remoteDownloader;

public synchronized void init(Downloader remoteDownloader) {
}